src/stdlib_str2num.fypp

```diff
@@ -0,0 +1,569 @@
+    integer(int8), parameter :: plus_sign  = ichar('+',int8) - digit_0
+    integer(int8), parameter :: Inf        = ichar('I',int8) 
+    integer(int8), parameter :: NaN        = ichar('N',int8) 
+    integer(int8), parameter :: le         = ichar('e',int8) - digit_0
+    integer(int8), parameter :: BE         = ichar('E',int8) - digit_0
+    integer(int8), parameter :: ld         = ichar('d',int8) - digit_0
+    integer(int8), parameter :: BD         = ichar('D',int8) - digit_0
+    integer(int8), parameter :: LF = 10, CR = 13, WS = 32
+
+    interface to_num
+        !! version: experimental
+        !!
+        !! Conversion of strings to numbers
+        !! ([Specification](../page/specs/stdlib_str2num.html#to-num-conversion-of-strings-to-numbers)) 
+        #:for k1, t1 in (INT_KINDS_TYPES + REAL_KINDS_TYPES)
+        module procedure to_${k1}$
+        #:endfor
+    end interface
+
+    interface to_num_p
+        !! version: experimental
+        !!
+        !! Conversion of a stream of values in a string to numbers
+        !! ([Specification](../page/specs/stdlib_str2num.html#to-num-p-conversion-of-a-stream-of-values-in-a-strings-to-numbers)) 
+        #:for k1, t1 in (INT_KINDS_TYPES + REAL_KINDS_TYPES)
+        module procedure to_${k1}$_p
+        #:endfor
+    end interface
+
+    interface to_num_base
+        #:for k1, t1 in (INT_KINDS_TYPES + REAL_KINDS_TYPES)
+        module procedure to_${k1}$_base
+        #:endfor
+    end interface
+    
+    contains
+    
+    !---------------------------------------------
+    ! String To Number interfaces
+    !---------------------------------------------
+
+    #:for k1, t1 in (INT_KINDS_TYPES + REAL_KINDS_TYPES)
+    elemental function to_${k1}$(s,mold) result(v)
+        ! -- In/out Variables
+        character(*), intent(in) :: s !> input string
+        ${t1}$, intent(in) :: mold !> dummy argument to disambiguate at compile time the generic interface
+        ${t1}$ :: v !> Output ${t1}$ value
+        ! -- Internal Variables
+        integer(int8) :: p !> position within the number
+        integer(int8)  :: stat !> error status
+        !----------------------------------------------
+        call to_num_base(s,v,p,stat)
+    end function
+    
+    function to_${k1}$_p(s,mold,stat) result(v)
+        ! -- In/out Variables
+        character(len=:), pointer :: s !> input string
+        ${t1}$, intent(in) :: mold !> dummy argument to disambiguate at compile time the generic interface
+        ${t1}$ :: v !> Output ${t1}$ value
+        integer(int8),intent(inout), optional :: stat
+        ! -- Internal Variables
+        integer(int8) :: p !> position within the number
+        integer(int8) :: err
+        !----------------------------------------------
+        call to_num_base(s,v,p,err)
+        p = min( p , len(s) )
+        s => s(p:)
+        if(present(stat)) stat = err
+    end function
+
+    #:endfor
```

Comment:
It would be nice to have an interface to a shallow wrapper that takes `string_type` as input. Better integration with the rest of stdlib.

Reply:
Haven't used the string_type from stdlib so didn't think about that. Maybe after merging the PR we could take a look at that?

Reply:
Definitely, since it's an additional feature, can be a future PR.